src/main/java/com/bmuschko/gradle/docker/internal/DefaultDockerUrlValueSource.java

```diff
@@ -0,0 +1,52 @@
+package com.bmuschko.gradle.docker.internal;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.provider.ValueSource;
+import org.gradle.api.provider.ValueSourceParameters;
+
+import javax.annotation.Nullable;
+import java.io.File;
+
+public abstract class DefaultDockerUrlValueSource implements ValueSource<String, ValueSourceParameters.None> {
```

Comment:
I don't think this class need to be `abstract` right? Only if you wanted to inject an object reference with `@Inject` I believe.

Reply:
There is a method in the interface that is documented in this way [in the Javadoc of ValueSource](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ValueSource.html):

> Do not implement this method in your subclass. Gradle provides the implementation when creating a provider from the value source via [ProviderFactory.of(Class, Action)](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ProviderFactory.html#of-java.lang.Class-org.gradle.api.Action-).

Reply:
I am still confused on when to use `abstract` or not on this class. IMHO this should be spelled out more explicitly by the documentation.

This is example uses it: https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:external_processes.
The Javadoc shows an example that doesn't: https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ValueSource.html

```
 public class MyValueSource implements ValueSource<...> {
     @Inject
     public MyValueSource(ExecOperations execOperations) {
         ...
     }
 }
```

Reply:
It must be abstract, and the Javadoc is incorrect and will be updated, thank you!
The getParameters() implementation must be provided by Gradle.

Reply:
Good deal.